Add CSS parser support for the @view-transition rule and the navigation descriptor.

mattwoodrow opened the pull request:
#### add19a6245520b35080caa82faff3ed001ac0caf
<pre>
Add CSS parser support for the @view-transition rule and the navigation descriptor.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277461">https://bugs.webkit.org/show_bug.cgi?id=277461</a>
&lt;<a href="https://rdar.apple.com/132947828">rdar://132947828</a>&gt;

Reviewed by Tim Nguyen.

Adds support for parsing the @view-transition rule, and the navigation
descriptor.

Adds the CSSViewTransitionRule.idl CSSOM class, and implementation.

Adds a partial implemention of &apos;6.2. Resolving the @view-transition rule&apos;
(<a href="https://www.w3.org/TR/css-view-transitions-2/#vt-rule-algo)">https://www.w3.org/TR/css-view-transitions-2/#vt-rule-algo)</a>, which still needs
support for transition types.

All of this is gated behind a new experimental preference
CSSCrossDocumentViewTransitionsEnabled.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-cssom-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSRule.idl:
* Source/WebCore/css/CSSViewTransitionRule.cpp: Added.
(WebCore::toViewTransitionNavigationEnum):
(WebCore::StyleRuleViewTransition::StyleRuleViewTransition):
(WebCore::StyleRuleViewTransition::create):
(WebCore::CSSViewTransitionRule::create):
(WebCore::CSSViewTransitionRule::CSSViewTransitionRule):
(WebCore::CSSViewTransitionRule::cssText const):
(WebCore::CSSViewTransitionRule::reattach):
* Source/WebCore/css/CSSViewTransitionRule.h: Added.
(isType):
* Source/WebCore/css/CSSViewTransitionRule.idl: Copied from Source/WebCore/css/StyleRuleType.h.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isViewTransitionRule const):
* Source/WebCore/css/StyleRuleType.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
(WebCore::StyleSheetContents::mayDependOnBaseURL const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::computeNewAllowedRules):
(WebCore::CSSParserImpl::consumeAtRule):
(WebCore::CSSParserImpl::consumeViewTransitionRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::CSSPropertyParser::parseViewTransitionDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::dispatchPagehideEvent):
(WebCore::Document::resolveViewTransitionRule):
* Source/WebCore/dom/Document.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::setViewTransitionRule):
(WebCore::Style::RuleSet::viewTransitionRule const):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::viewTransitionRule const):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::viewTransitionRule const):
* Source/WebCore/style/StyleScopeRuleSets.h:

Canonical link: <a href="https://commits.webkit.org/281694@main">https://commits.webkit.org/281694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2703940df4ce18cf5ec74d145b01e2173de7b902

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40077 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13294 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64649 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11265 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11540 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49099 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7813 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62752 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37324 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52598 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29930 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34012 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9830 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10178 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/53818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66378 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59965 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9960 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56463 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4683 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52572 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/56643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13529 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3863 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81720 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14223 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36964 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/38057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSViewTransitionRule.cpp:
```diff
@@ -0,0 +1,102 @@
+        builder.append("; "_s);
+    }
+
+    builder.append("}"_s);
```
```suggestion
    builder.append('}');
```

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -984,6 +984,20 @@ CSSCounterStyleAtRulesEnabled:
+    WebKit:
+      default: false
+    WebCore:
+      default: false
```
nit: maybe this, to be consistent with the VT flag, and because there's more than just CSS in this bucket:

```suggestion
CrossDocumentViewTransitionsEnabled:
  type: bool
  status: testable
  category: animation
  humanReadableName: "Cross-document View Transitions"
  humanReadableDescription: "Enable support for cross-document view transitions"
  defaultValue:
    WebKitLegacy:
      default: false
    WebKit:
      default: false
    WebCore:
      default: false
```

nt1m commented on Source/WebCore/css/CSSRule.idl:
```diff
@@ -49,4 +49,5 @@
     [ImplementedAs=CounterStyle, EnabledBySetting=CSSCounterStyleAtRulesEnabled] const unsigned short COUNTER_STYLE_RULE = 11;
     [ImplementedAs=Supports] const unsigned short SUPPORTS_RULE = 12;
     [ImplementedAs=FontFeatureValues] const unsigned short FONT_FEATURE_VALUES_RULE = 14;
+    [ImplementedAs=ViewTransition, EnabledBySetting=CSSCrossDocumentViewTransitionsEnabled] const unsigned short VIEW_TRANSITION_RULE = 15;
```
I removed this from the spec: https://github.com/w3c/csswg-drafts/commit/f1d42e606113d7a7e7f45fcbd9dad1bd6ce9e775

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -4778,6 +4778,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
     case CSSPropertySizeAdjust:
         return nullptr;
 
+    // Unimplemented @view-transition properties.
```
```suggestion
    // @view-transition descriptors.
```

nit: Probably more accurate to say they're not supported because they're descriptors.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8067,6 +8068,17 @@ void Document::dispatchPagehideEvent(PageshowEventPersistence persisted)
     dispatchWindowEvent(PageTransitionEvent::create(eventNames().pagehideEvent, persisted == PageshowEventPersistence::Persisted), this);
 }
 
+bool Document::resolveViewTransitionRule()
```
```suggestion
bool Document::shouldPerformCrossDocumentViewTransition()
```

I'm not sure what you're planning to use this for, but the name could be more descriptive.

(Maybe that should take an `origin` parameter and check if it's the same origin too)

nt1m replied:
Matt pointed out this matches a spec algorithm, this is fine as it is, with a spec link added.

mattwoodrow committed:
Add CSS parser support for the @view-transition rule and the navigation descriptor.
https://bugs.webkit.org/show_bug.cgi?id=277461
<rdar://132947828>

Reviewed by Tim Nguyen.

Adds support for parsing the @view-transition rule, and the navigation
descriptor.

Adds the CSSViewTransitionRule.idl CSSOM class, and implementation.

Adds a partial implemention of '6.2. Resolving the @view-transition rule'
(https://www.w3.org/TR/css-view-transitions-2/#vt-rule-algo), which still needs
support for transition types.

All of this is gated behind a new experimental preference
CSSCrossDocumentViewTransitionsEnabled.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-cssom-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSRule.idl:
* Source/WebCore/css/CSSViewTransitionRule.cpp: Added.
(WebCore::toViewTransitionNavigationEnum):
(WebCore::StyleRuleViewTransition::StyleRuleViewTransition):
(WebCore::StyleRuleViewTransition::create):
(WebCore::CSSViewTransitionRule::create):
(WebCore::CSSViewTransitionRule::CSSViewTransitionRule):
(WebCore::CSSViewTransitionRule::cssText const):
(WebCore::CSSViewTransitionRule::reattach):
* Source/WebCore/css/CSSViewTransitionRule.h: Added.
(isType):
* Source/WebCore/css/CSSViewTransitionRule.idl: Copied from Source/WebCore/css/StyleRuleType.h.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isViewTransitionRule const):
* Source/WebCore/css/StyleRuleType.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
(WebCore::StyleSheetContents::mayDependOnBaseURL const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::computeNewAllowedRules):
(WebCore::CSSParserImpl::consumeAtRule):
(WebCore::CSSParserImpl::consumeViewTransitionRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue):
(WebCore::CSSPropertyParser::parseViewTransitionDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::dispatchPagehideEvent):
(WebCore::Document::resolveViewTransitionRule):
* Source/WebCore/dom/Document.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::setViewTransitionRule):
(WebCore::Style::RuleSet::viewTransitionRule const):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::viewTransitionRule const):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::viewTransitionRule const):
* Source/WebCore/style/StyleScopeRuleSets.h:

Canonical link: https://commits.webkit.org/281694@main

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.h:
```diff
@@ -56,6 +56,7 @@ class CSSPropertyParser {
     static bool isValidCustomPropertyValueForSyntax(const CSSCustomPropertySyntax&, CSSParserTokenRange, const CSSParserContext&);
 
     static RefPtr<CSSValue> parseCounterStyleDescriptor(CSSPropertyID, CSSParserTokenRange&, const CSSParserContext&);
+    static RefPtr<CSSValue> parseViewTransitionDescriptor(CSSPropertyID, CSSParserTokenRange&, const CSSParserContext&);
```
This method is unused, can you remove it?
